third_party/move/move-vm/runtime/src/storage/environment.rs

```diff
@@ -0,0 +1,266 @@
+    pub fn vm_config(&self) -> &VMConfig {
+        &self.vm_config
+    }
+
+    /// Returns native functions available to this runtime.
+    pub(crate) fn natives(&self) -> &NativeFunctions {
+        &self.natives
+    }
+
+    /// Returns the re-indexing map currently used by this runtime environment
+    /// to remap struct identifiers into indices.
+    pub(crate) fn struct_name_index_map(&self) -> &StructNameIndexMap {
+        &self.struct_name_index_map
+    }
+
+    /// Returns the cloned environment, with a deep-clone of struct name index map.
+    pub fn clone_with_new_struct_name_index_map(&self) -> Self {
+        Self {
+            vm_config: self.vm_config.clone(),
+            natives: self.natives.clone(),
+            verifier_extension: self.verifier_extension.clone(),
+            struct_name_index_map: self.struct_name_index_map.clone(),
+        }
+    }
+
+    /// Creates a partially verified compiled script by running:
+    ///   1. Move bytecode verifier,
+    ///   2. Verifier extension, if provided.
+    pub fn build_partially_verified_script(
+        &self,
+        compiled_script: Arc<CompiledScript>,
+    ) -> VMResult<PartiallyVerifiedScript> {
+        move_bytecode_verifier::verify_script_with_config(
+            &self.vm_config().verifier_config,
+            compiled_script.as_ref(),
+        )?;
+        if let Some(verifier) = &self.verifier_extension {
+            verifier.verify_script(compiled_script.as_ref())?;
+        }
+        Ok(PartiallyVerifiedScript(compiled_script))
+    }
+
+    /// Creates a fully verified script by running dependency verification
+    /// pass. The caller must provide verified module dependencies.
+    pub fn build_verified_script(
+        &self,
+        partially_verified_script: PartiallyVerifiedScript,
+        immediate_dependencies: &[Arc<Module>],
+    ) -> VMResult<Script> {
+        dependencies::verify_script(
+            partially_verified_script.0.as_ref(),
+            immediate_dependencies.iter().map(|m| m.module()),
+        )?;
+        Script::new(partially_verified_script.0, self.struct_name_index_map())
+            .map_err(|e| e.finish(Location::Script))
+    }
+
+    /// Creates a partially verified compiled module by running:
+    ///   1. Move bytecode verifier,
+    ///   2. Verifier extension, if provided.
+    pub fn build_partially_verified_module(
+        &self,
+        compiled_module: Arc<CompiledModule>,
+        module_size: usize,
+        module_hash: &[u8; 32],
+    ) -> VMResult<PartiallyVerifiedModule> {
+        if !VERIFIED_MODULES_V2.contains(module_hash) {
+            // For regular execution, we cache already verified modules. Note
+            // that this even caches verification for the published modules.
+            // This should be ok because as long as the hash is the same, the
+            // deployed bytecode and any dependencies are the same, and so the
+            // cached verification result can be used.
```

Comment:
is this true? module hash doesn't include information about dependencies iiuc. also the verification result could change with bytecode verifier version/config update

Reply:
Hash includes all binary, so even dependency handles? Here verification is local, so we do not know if this module is ok wrt dependencies, but locally same hash should guarantee the equivalence?

This does not work with changes in verifier, but current code also does not work with that iirc @runtian-zhou? Ideally, we avoid this cache by having a single cache outside of the block which captures verification per epoch.

Added it to compare to V1 loader because if it is not there, V1 is faster :(

Reply:
hash includes dependency identifier but not dependency content right? i assume compatibility rules enforce once it's published, dependency upgrade can't break it but want to double check